activiti-spring-process-extensions/src/main/java/org/activiti/spring/process/ProcessVariablesInitiator.java

```diff
@@ -0,0 +1,101 @@
+/* Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.activiti.spring.process;
+
+import org.activiti.bpmn.model.FlowElement;
+import org.activiti.bpmn.model.Process;
+import org.activiti.engine.ActivitiException;
+import org.activiti.engine.impl.util.ProcessInstanceHelper;
+import org.activiti.engine.repository.ProcessDefinition;
+import org.activiti.engine.runtime.ProcessInstance;
+import org.activiti.spring.process.model.ProcessExtensionModel;
+import org.activiti.spring.process.model.VariableDefinition;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.text.ParseException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class ProcessVariablesInitiator extends ProcessInstanceHelper {
```

Comment:
@balsarori @ryandawsonuk this class doesn't make any sense. If we want to modify the behaviour of ProcessInstanceHelper we should do so to support process variables definitions. This class hierarchy will cause us a lot of trouble so this is definitely not the way to go. We need to change this.

Reply:
What I suggest is to push this logic to ProcessInstanceHelper, which will push us to make ProcessInstanceHelper more flexible and generic.

Reply:
@ryandawsonuk I still dislike the inheritance here. Unless we are planing to haa ProcessInstanceHelper for Cloud, having more fine grained versions of it doesn't make too much sense. The name "ProcessInstanceHelperForExtensions" implies that it will handle only Extensions, which means that if we want to add any other customization for cloud/spring boot the name will become a problem. 
@balsarori even if the engine doesn't need to know, the new APIs will need to know, so we just need to make sure that we make it available at the right level and with the right name.

Reply:
@salaboy  APIs don't cover all possible ways of starting a new process instance. I don't like the inheritance either so I would suggest that we have allow the ability to register process instance initiators.

Reply:
@balsarori "APIs don't cover all possible ways of starting a new process instance." we need to make sure that the supported ways would work without any extra configuration. Right now we are not interested in covering all the possible ways, we will add more ways when the basics are more stable. 
Adding more than ProcessInstanceInitiator will come with more challenges such as order and priority. If by register you mean to select which one you would like then yes.. that is more being able to configure which one to use.

Reply:
> "we need to make sure that the supported ways would work without any extra configuration" 

that's not possible. I don't understand why shouldn't we cover all possible process starts if there is an easy hook that we can add to the engine to do that.

> Adding more than ProcessInstanceInitiator will come with more challenges such as order and priority. 

Why order? I don't think that should be a concern until a need for that arises. 

I guess we are over complicating things while they are straight forward and easy to implement. Your concerns might or might not be valid however we can always refactor if things prove to not be right later.